### PR TITLE
Fix ViewTarget is garbage collected during actions tracking

### PR DESCRIPTION
**Pull request by ambushwork**
### What does this PR do?

Prior to this PR, we were using `WeakReference` on `ViewTarget`, since `ViewTarget` is a local variable only in the scope of `findTarget` function, after execution, there will be zero reference of it, and if GC occurs, the object in  `WeakReference` will be cleaned. 

The correct way is that we should use strong reference of `ViewTarget`, but inside `ViewTarget` we hold weakReference of `View`. 

Also a dedicated GC unit test is added to cover this case.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: features/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListenerScrollSwipeTest.kt [hunk 1]
```diff
@@ -280,6 +280,79 @@ internal class GesturesListenerScrollSwipeTest : AbstractGesturesListenerTest()
...
+            internalLogger = mockInternalLogger
+        )
+
+        // When
+        testedListener.onDown(startDownEvent)
+        stubStopMotionEvent(endUpEvent, startDownEvent, expectedDirection1)
+        intermediaryEvents.forEachIndexed { index, event ->
+            testedListener.onScroll(startDownEvent, event, distancesX[index], distancesY[index])
+        }
+        System.gc()
```
Comment (jonathanmos) [inline features/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListenerScrollSwipeTest.kt:343]: System.gc can't guarantee that gc will occur, so my suspicion was that this line would make the test flaky. However even after removing this line the test seems to pass every time, which seems strange to me

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt [hunk 2]
```diff
@@ -7,18 +7,41 @@
...
+    val viewRef: WeakReference<View?> = WeakReference(null),
     val tag: String? = null
-)
+) {
+
+    override fun equals(other: Any?): Boolean {
+        // Overriding hashcode & equals because we should compare the referent
+        // instead of the reference.
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
```
Comment (satween) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt:32]: Suggestions: 

```
if (other !is ViewTarget) return false 
``` 
let you remove `other as ViewTarget` and ```if (javaClass != other?.javaClass)```

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt [hunk 3]
```diff
@@ -7,18 +7,41 @@
...
+
+        if (viewRef.get() != other.viewRef.get()) return false
+        if (tag != other.tag) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = viewRef.get().hashCode()
+        result = 31 * result + (tag?.hashCode() ?: 0)
```
Comment (satween) [inline features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt:44]: You can call ```tag.hashCode()``` (without ? mark) here. Extension function does literally the same ( returns 0 if null)

File: features/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListenerScrollSwipeTest.kt (hunk 1 above)
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListenerScrollSwipeTest.kt:343]: you are right about removing the line the test still pass, it is because this unit test is to protect against a corner case which is "gc happened during the scrolling", after removing the line, it changes from a corner case to a normal case "nothing happened during the scrolling", so it will pass for sure.

But why in my opinion it is not a false positive test? I wrote this test before I made the fix in the production code, and it will fail 100%, after making the fix, the unit test can pass.

and yes, when calling `system.gc()` may not guarantee gc occurs ( although very unlikely in test environment from what I see), it doesn't fail the test, it just rollback to a normal case that I mentioned above.

In this case I don't see it as flaky one, what do you think?

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt [hunk 4]
```diff
@@ -7,18 +7,41 @@
...
+    override fun equals(other: Any?): Boolean {
+        // Overriding hashcode & equals because we should compare the referent
+        // instead of the reference.
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+
+        other as ViewTarget
+
+        if (viewRef.get() != other.viewRef.get()) return false
+        if (tag != other.tag) return false
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt:37]: it's expected, if the `viewRefs` of two `ViewTarget` are the same but tags are different,  then they should be seen as different object.

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt [hunk 5]
```diff
@@ -7,18 +7,41 @@
...
+ * Only one of [viewRef] or [tag] is expected to be non-null, depending on the UI framework used.
  */
 data class ViewTarget(
-    val view: View? = null,
+    val viewRef: WeakReference<View?> = WeakReference(null),
     val tag: String? = null
-)
+) {
+
+    override fun equals(other: Any?): Boolean {
```
[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt:28]: updated with the data class test

**Review by jonathanmos — DISMISSED** [review]

**Review by mariusc83 — APPROVED** [review]

File: features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt:35]: I'm wondering if it is useful to `ViewTarget` as `data` class. What benefit does it give?

Is it `data` class just for the assertion in the RUM configuration related tests?

[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt:35]: I used `data` class just to have the convenience for the comparison of object at the beginning, now I overwrite `hashcode()` and `equals()` it has no use anymore, I can remove it

[the remark this answers is not in the document]
Comment (0xnm) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt:35]: I mean do we even need to overwrite `hashcode` and `equals` if `ViewTarget` is not used as a key or for the comparison?

[the remark this answers is not in the document]
Comment (ambushwork) [reply features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/tracking/ViewTarget.kt:35]: we are comparing it here:
https://github.com/DataDog/dd-sdk-android/blob/yl/fix-action-tracking-flaky/features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/instrumentation/gestures/GesturesListener.kt#L216
